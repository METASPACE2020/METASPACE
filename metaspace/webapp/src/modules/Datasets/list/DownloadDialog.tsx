import { computed, createComponent, toRefs, watch } from '@vue/composition-api'
import { useQuery } from '@vue/apollo-composable'
import { DownloadLinkJson, GetDatasetDownloadLink, getDatasetDownloadLink } from '../../../api/dataset'
import safeJsonParse from '../../../lib/safeJsonParse'
import { Dialog } from 'element-ui'
import router from '../../../router'

const getExtension = (filename: string) => {
  const lastDot = filename.lastIndexOf('.')
  if (lastDot !== -1) {
    const ext = filename.slice(lastDot)
    if (ext.toLowerCase() === '.imzml') {
      return '.imzML'
    } else {
      return ext.toLowerCase()
    }
  }
  return ''
}

const FileIcon = createComponent<{filename: string}>({
  props: { filename: { type: String, required: true } },
  setup(props) {
    return () => (
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 72 100" style="width: 32px; height: 50px">
        <linearGradient id="a" gradientUnits="userSpaceOnUse" x1="36" y1="99" x2="36" y2="1">
          <stop offset="0" stop-color="#c8d4db"/>
          <stop offset=".139" stop-color="#d8e1e6"/>
          <stop offset=".359" stop-color="#ebf0f3"/>
          <stop offset=".617" stop-color="#f9fafb"/>
          <stop offset="1" stop-color="#fff"/>
        </linearGradient>
        <path d="M45 1L71 27.7V99H1V1H45Z" fill="url(#a)" stroke="#7191a1" stroke-width="2"/>
        <linearGradient id="b" gradientUnits="userSpaceOnUse" x1="45.068" y1="27.796" x2="58.568" y2="14.395">
          <stop offset="0" stop-color="#fff"/>
          <stop offset=".35" stop-color="#fafbfb"/>
          <stop offset=".532" stop-color="#edf1f4"/>
          <stop offset=".675" stop-color="#dde5e9"/>
          <stop offset=".799" stop-color="#c7d3da"/>
          <stop offset=".908" stop-color="#adbdc7"/>
          <stop offset="1" stop-color="#92a5b0"/>
        </linearGradient>
        <path d="M45 1L71 27.7H45V1z" fill="url(#b)"/><path d="M45 1l27 26.7H45V1z" fill-opacity="0" stroke="#7191a1" stroke-width="2" stroke-linejoin="bevel"/>
        <text x="36" y="60" text-length="64" text-anchor="middle" font-size="20px" font-weight="bold" fill="#7191a1">{getExtension(props.filename)}</text>
      </svg>
    )
  },
})

export default createComponent({
  props: {
    datasetName: { type: String, required: true },
    datasetId: { type: String, required: true },
  },
  setup(props, { emit }) {
    const { datasetId } = toRefs(props)
    const {
      result: downloadLinkResult,
      loading,
    } = useQuery<GetDatasetDownloadLink>(getDatasetDownloadLink, { datasetId }, { fetchPolicy: 'no-cache' })
    const downloadLinks = computed<DownloadLinkJson>(() => downloadLinkResult.value != null
      && safeJsonParse(downloadLinkResult.value.dataset.downloadLinkJson))

    const datasetShareHref = computed(() => router.resolve({
      path: '/datasets',
      query: { ds: datasetId.value },
    }).href)

    return () => {
      let content
      if (loading.value) {
        content = <div v-loading style='min-height: 500px' />
      } else if (downloadLinks.value == null) {
        content = <div><h1>Error</h1><p>This dataset cannot be downloaded.</p></div>
      } else {
        const { license, contributors, files } = downloadLinks.value
        content = (
          <div>
            {license.code === 'NO-LICENSE'
              ? <p>
                This dataset's submitter has not selected a license.
                Please get approval from the submitter before downloading or using this dataset in any way.
              </p>
              : <p>
                This dataset is distributed under the <a href={license.link}>{license.name}</a>.
                Please check its terms before you download or use this dataset.
              </p>}

            <p>
              Make sure to credit the {contributors.length !== 1 ? 'authors' : 'author'} of this dataset:
            </p>
            <ul>
              {contributors.map(({ name, institution }) =>
                <li>{name}{institution && ', ' + institution}</li>)}
            </ul>
            <h4>Files</h4>
            <div>
              {files.map(({ filename, link }) => (
                <div style="display: flex; align-items: center; padding: 5px; margin: 5px 0;">
                  <a href={link}>
                    <FileIcon filename={getExtension(filename)} />
                  </a>
                  <a href={link} style="margin: 5px 10px">
                    {filename}
                  </a>
                </div>
              ))}
            </div>
            <p>
              <i>
              These download links expire after 30 minutes.
              </i>
            </p>
          </div>
        )
      }
      (window as any).content = content

      return (
        <Dialog
          visible
          lockScroll={false}
          onclose={() => emit('close')}
          title={`Download ${props.datasetName}`}
        >
          {content.children}
        </Dialog>
      )
    }
  },
})
