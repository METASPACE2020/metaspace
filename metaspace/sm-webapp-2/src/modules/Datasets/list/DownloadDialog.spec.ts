import { flushPromises, mount } from '@vue/test-utils'
import { defineComponent, nextTick, ref, h } from 'vue'
import DownloadDialog from './DownloadDialog'
import { vi } from 'vitest'

const TestDownloadDialog = defineComponent({
  name: 'DownloadDialogWrapper',
  components: {
    DownloadDialog,
  },
  props: ['datasetId', 'datasetName'],
  setup(props) {
    return () => h(DownloadDialog, { ...props })
  },
})

describe('Datasets/DownloadDialog', () => {
  it('should match snapshot', async () => {
    vi.mock('@vue/apollo-composable', () => ({
      useQuery: vi.fn(() => ({
        result: ref({
          dataset: {
            downloadLinkJson: JSON.stringify({
              contributors: [{ name: 'Foo Bar', institution: 'EMBL' }],
              license: {
                code: 'CC BY 4.0',
                name: 'Creative Commons Attribution 4.0 International Public License',
                link: 'https://creativecommons.org/licenses/by/4.0/',
              },
              files: [
                {
                  filename: 'Image5.ibd',
                  link:
                    'https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test' +
                    '&Expires=1585569522&Signature=test%3D',
                },
                {
                  filename: 'Image5.imzML',
                  link:
                    'https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test' +
                    '&Expires=1585569522&Signature=test%3D',
                },
              ],
            }),
          },
        }),
        loading: ref(false),
        error: ref(null),
      })),
    }))

    const propsData = {
      datasetId: '2020-01-02_03h04m05s',
      datasetName: 'Mouse whole body DHB',
    }

    const wrapper = mount(TestDownloadDialog, {
      propsData,
    })

    await flushPromises()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
