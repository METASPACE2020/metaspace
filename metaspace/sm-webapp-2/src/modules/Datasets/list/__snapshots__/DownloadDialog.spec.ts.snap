// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Datasets/DownloadDialog > should match snapshot 1`] = `
"<div class=\\"mock-el-dialog\\" model-value=\\"true\\" lockscroll=\\"false\\" title=\\"Download Mouse whole body DHB\\">
  <p>
  <p>This dataset is distributed under the <a href=\\"https://creativecommons.org/licenses/by/4.0/\\" rel=\\"noopener noreferrer nofollow\\" target=\\"_blank\\">CC BY 4.0</a> license. Make sure to credit the author of this dataset:</p>
  </p>
  <ul>
    <li class=\\"list-none\\">Foo Bar, EMBL</li>
  </ul>
  <h4>Files</h4>
  <div class=\\"flex my-1 flex-wrap\\">
    <div class=\\"flex items-center p-1 w-1/2 box-border\\"><a href=\\"https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test&amp;Expires=1585569522&amp;Signature=test%3D\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 72 100\\" style=\\"width: 32px; height: 50px;\\">
          <linearGradient id=\\"a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"36\\" y1=\\"99\\" x2=\\"36\\" y2=\\"1\\">
            <stop offset=\\"0\\" stop-color=\\"#c8d4db\\"></stop>
            <stop offset=\\".139\\" stop-color=\\"#d8e1e6\\"></stop>
            <stop offset=\\".359\\" stop-color=\\"#ebf0f3\\"></stop>
            <stop offset=\\".617\\" stop-color=\\"#f9fafb\\"></stop>
            <stop offset=\\"1\\" stop-color=\\"#fff\\"></stop>
          </linearGradient>
          <path d=\\"M45 1L71 27.7V99H1V1H45Z\\" fill=\\"url(#a)\\" stroke=\\"#7191a1\\" stroke-width=\\"2\\"></path>
          <linearGradient id=\\"b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.068\\" y1=\\"27.796\\" x2=\\"58.568\\" y2=\\"14.395\\">
            <stop offset=\\"0\\" stop-color=\\"#fff\\"></stop>
            <stop offset=\\".35\\" stop-color=\\"#fafbfb\\"></stop>
            <stop offset=\\".532\\" stop-color=\\"#edf1f4\\"></stop>
            <stop offset=\\".675\\" stop-color=\\"#dde5e9\\"></stop>
            <stop offset=\\".799\\" stop-color=\\"#c7d3da\\"></stop>
            <stop offset=\\".908\\" stop-color=\\"#adbdc7\\"></stop>
            <stop offset=\\"1\\" stop-color=\\"#92a5b0\\"></stop>
          </linearGradient>
          <path d=\\"M45 1L71 27.7H45V1z\\" fill=\\"url(#b)\\" stroke-width=\\"2\\" stroke=\\"#7191a1\\" stroke-linejoin=\\"bevel\\"></path><text x=\\"36\\" y=\\"60\\" text-length=\\"64\\" text-anchor=\\"middle\\" font-size=\\"20px\\" font-weight=\\"bold\\" fill=\\"#7191a1\\">.ibd</text>
        </svg></a><a href=\\"https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test&amp;Expires=1585569522&amp;Signature=test%3D\\" class=\\"flex mx-3 my-1 overflow-hidden\\"><span class=\\"flex-shrink truncate\\">Image5</span><span>.ibd</span></a></div>
    <div class=\\"flex items-center p-1 w-1/2 box-border\\"><a href=\\"https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test&amp;Expires=1585569522&amp;Signature=test%3D\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 72 100\\" style=\\"width: 32px; height: 50px;\\">
          <linearGradient id=\\"a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"36\\" y1=\\"99\\" x2=\\"36\\" y2=\\"1\\">
            <stop offset=\\"0\\" stop-color=\\"#c8d4db\\"></stop>
            <stop offset=\\".139\\" stop-color=\\"#d8e1e6\\"></stop>
            <stop offset=\\".359\\" stop-color=\\"#ebf0f3\\"></stop>
            <stop offset=\\".617\\" stop-color=\\"#f9fafb\\"></stop>
            <stop offset=\\"1\\" stop-color=\\"#fff\\"></stop>
          </linearGradient>
          <path d=\\"M45 1L71 27.7V99H1V1H45Z\\" fill=\\"url(#a)\\" stroke=\\"#7191a1\\" stroke-width=\\"2\\"></path>
          <linearGradient id=\\"b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.068\\" y1=\\"27.796\\" x2=\\"58.568\\" y2=\\"14.395\\">
            <stop offset=\\"0\\" stop-color=\\"#fff\\"></stop>
            <stop offset=\\".35\\" stop-color=\\"#fafbfb\\"></stop>
            <stop offset=\\".532\\" stop-color=\\"#edf1f4\\"></stop>
            <stop offset=\\".675\\" stop-color=\\"#dde5e9\\"></stop>
            <stop offset=\\".799\\" stop-color=\\"#c7d3da\\"></stop>
            <stop offset=\\".908\\" stop-color=\\"#adbdc7\\"></stop>
            <stop offset=\\"1\\" stop-color=\\"#92a5b0\\"></stop>
          </linearGradient>
          <path d=\\"M45 1L71 27.7H45V1z\\" fill=\\"url(#b)\\" stroke-width=\\"2\\" stroke=\\"#7191a1\\" stroke-linejoin=\\"bevel\\"></path><text x=\\"36\\" y=\\"60\\" text-length=\\"64\\" text-anchor=\\"middle\\" font-size=\\"20px\\" font-weight=\\"bold\\" fill=\\"#7191a1\\">.imzML</text>
        </svg></a><a href=\\"https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test&amp;Expires=1585569522&amp;Signature=test%3D\\" class=\\"flex mx-3 my-1 overflow-hidden\\"><span class=\\"flex-shrink truncate\\">Image5</span><span>.imzML</span></a></div>
  </div>
  <p><i>These download links expire after 30 minutes.</i></p>
</div>"
`;
